io/src/main/scala/sbt/internal/io/Milli.scala

```diff
@@ -319,11 +319,19 @@ private abstract class MilliMilliseconds extends Milli {
 
 private object JavaMilli extends MilliMilliseconds {
   def getModifiedTime(filePath: String): Long =
-    Files.getLastModifiedTime(JPaths.get(filePath)).toMillis
-  def setModifiedTime(filePath: String, mtime: Long): Unit = {
-    Files.setLastModifiedTime(JPaths.get(filePath), FileTime.fromMillis(mtime))
-    ()
-  }
+    mapNoSuchFileException(Files.getLastModifiedTime(JPaths.get(filePath)).toMillis)
+  def setModifiedTime(filePath: String, mtime: Long): Unit =
+    mapNoSuchFileException {
+      Files.setLastModifiedTime(JPaths.get(filePath), FileTime.fromMillis(mtime))
+      ()
+    }
+
+  private def mapNoSuchFileException[A](f: => A): A =
+    try {
+      f
+    } catch {
+      case e: NoSuchFileException => throw new FileNotFoundException(e.getFile)
```

Comment:
An issue here is that the original stack trace is lost. It is not a big problem since `mapNoSuchFileException()` is only used once, but if it were used from more than one method, it would not be clear which code path led to the exception. Normally that can be avoided by chaining exceptions; however, it just so happens that `FileNotFoundException` does not have a `Throwable`-based constructor, so chaining is a bit more difficult than usual. Still, there is a workaround:
```
case e: NoSuchFileException => throw new FileNotFoundException(e.getFile) {
  initCause(e)
  override def toString() = {
    val msg = getLocalizedMessage()
    getClass().getSuperclass().getName() + (if (msg == null) "" else (": " + msg))
  }
}
```
This will be reported as:
```
java.io.FileNotFoundException: filename
        at ...
Caused by: java.nio.file.NoSuchFileException: filename
        at ...
```
Where the second exception identifies the original stacktrace. The additional `toString()` is needed in order to keep the exception identifying itself as a `FileNotFoundException` in the printout.

Reply:
I think you can just call [`initCause`](https://docs.oracle.com/javase/8/docs/api/java/lang/Throwable.html#initCause-java.lang.Throwable-).

Reply:
As you prefer.